From df41e359fffce571d1b94b5c40e5462e0a68e7c5 Mon Sep 17 00:00:00 2001
Message-Id: <df41e359fffce571d1b94b5c40e5462e0a68e7c5.1415496071.git.chang-joon.lee@intel.com>
In-Reply-To: <aa0c58fe046f9d00f3d381a75f6169b3f6d58c94.1415496071.git.chang-joon.lee@intel.com>
References: <aa0c58fe046f9d00f3d381a75f6169b3f6d58c94.1415496071.git.chang-joon.lee@intel.com>
From: Brad Volkin <bradley.d.volkin@intel.com>
Date: Tue, 1 Jul 2014 11:05:02 -0700
Subject: [PATCH 6/8] FOR_UPSTREAM [VPG]: drm/i915: Switch command parser to
 enabling mode

Without this change, the parser is in 'enforcing' mode, where it will
validate batches but leave hardware parsing enabled. This means that
whatever additional functionality the software parser would allow is
still disallowed by hardware.

There is currently an upstream regression when full PPGTT is enabled
which makes this change unsafe in that case (we hit a BUG_ON a few
lines down). So for now, only switch to enabling mode when full PPGTT
is disabled, which should be the case for VLV.

For: APDEV-870
Change-Id: I56b23dbb3390d0a12748d392fc8e9163c1c188ce
Signed-off-by: Brad Volkin <bradley.d.volkin@intel.com>
---
 drivers/gpu/drm/i915/i915_gem_execbuffer.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_gem_execbuffer.c b/drivers/gpu/drm/i915/i915_gem_execbuffer.c
index 06e8424..fc452d7 100644
--- a/drivers/gpu/drm/i915/i915_gem_execbuffer.c
+++ b/drivers/gpu/drm/i915/i915_gem_execbuffer.c
@@ -1507,13 +1507,18 @@ i915_gem_do_execbuffer(struct drm_device *dev, void *data,
 			goto err;
 
 		/*
-		 * XXX: Actually do this when enabling batch copy...
+		 * XXX: Actually do this when enabling batch copy...and the
+		 * full PPGTT secure batch regression is fixed.
 		 *
 		 * Set the DISPATCH_SECURE bit to remove the NON_SECURE bit
 		 * from MI_BATCH_BUFFER_START commands issued in the
 		 * dispatch_execbuffer implementations. We specifically don't
 		 * want that set when the command parser is enabled.
 		 */
+		if (!USES_FULL_PPGTT(dev))
+			flags |= I915_DISPATCH_SECURE;
+		else
+			pr_err_once("CMD: trying to use command parser with full PPGTT\n");
 	}
 
 	/* snb/ivb/vlv conflate the "batch in ppgtt" bit with the "non-secure
-- 
1.7.9.5

